test/system/src/test/java/io/pravega/test/system/SegmentReaderTest.java

```diff
@@ -0,0 +1,174 @@
+    }
+
+    @Before
+    public void setup() {
+        Service conService = Utils.createPravegaControllerService(null);
+        List<URI> ctlURIs = conService.getServiceDetails();
+        controllerURI = ctlURIs.get(0);
+
+        streamManager = StreamManager.create(Utils.buildClientConfig(controllerURI));
+        assertTrue("Creating scope", streamManager.createScope(SCOPE));
+        assertTrue("Creating stream", streamManager.createStream(SCOPE, STREAM, config));
+    }
+
+    @After
+    public void tearDown() {
+        streamManager.close();
+        ExecutorServiceHelpers.shutdown(executor);
+    }
+
+    /**
+     * This test verifies the basic functionality of {@link SegmentReaderManager}, including fetching segment readers for
+     * stream, parallel segment reads and segment reader snapshot check.
+     */
+    @Test
+    public void segmentReaderTest() {
+        final int totalEvents = 500;
+        final Stream stream = Stream.of(SCOPE, STREAM);
+        final ClientConfig clientConfig = Utils.buildClientConfig(controllerURI);
+
+        @Cleanup
+        ConnectionFactory connectionFactory = new SocketConnectionFactoryImpl(clientConfig);
+        ControllerImpl controller = new ControllerImpl(ControllerImplConfig.builder().clientConfig(clientConfig).build(),
+                connectionFactory.getInternalExecutor());
+        @Cleanup
+        ClientFactoryImpl clientFactory = new ClientFactoryImpl(SCOPE, controller, connectionFactory);
+        @Cleanup
+        SegmentReaderManager<String> segmentReaderManager = SegmentReaderManager.create(clientConfig, new JavaSerializer<>());
+        log.info("Invoking segmentReader test with Controller URI: {}", controllerURI);
+
+        List<SegmentReader<String>> segmentReaderList = segmentReaderManager.getSegmentReaders(stream, null).join();
+
+        log.info("Writing events to stream {}.", stream);
+        // Write events to the Stream.
+        writeEvents(clientFactory, STREAM, totalEvents);
+        log.info("Writing events to stream {} finished.", stream);
+
+        log.debug("Sealing stream {}.", stream);
+        streamManager.sealStream(SCOPE, STREAM);
+
+        log.info("Reading events from stream {}.", stream);
+        int totalEventRead = readFromSegmentReaders(segmentReaderList);
+
+        assertEquals("Expected events read", totalEventRead, totalEvents);
+
+    }
+
+    // Start utils region
+
+    private int readFromSegmentReaders(List<SegmentReader<String>> segmentReaders) {
+        return segmentReaders.parallelStream()
+                .map(reader -> CompletableFuture.supplyAsync(() -> readEvents(reader))).collect(Collectors.toList())
+                .stream().map(CompletableFuture::join).mapToInt(Integer::intValue).sum();
+    }
+
+    private int readEvents(SegmentReader<String> segmentReader) {
+        int eventRead = 0;
+
+        while (true) {
+            try {
+                segmentReader.read(1000);
+                eventRead++;
```

Comment:
This increments eventRead even if the read times out. To prevent assertion failures in the caller it should probably only increment this if an event was returned.

Reply:
Done.